aip/0194.md

```diff
@@ -0,0 +1,237 @@
+
+### Non-retryable codes
+
+These codes **should** never be retried for any request:
+
+- `OK`: The request succeeded.
+- `CANCELLED`: A user can cancel a request, which **must** be honoured.
+- `DEADLINE_EXCEEDED`: A user can set a deadline, which **must** be honoured.
+- `INVALID_ARGUMENT`: Retrying a request with an invalid argument will never
+  succeed.
+- `DATA_LOSS`: This is an unrecoverable error and **must** immediately be
+  surfaced to the user.
+
+### Generally non-retryable codes:
+
+Some codes can be retried if a system is designed without synchronization
+between various components. For example, a system in which a read is designed to
+hang forever until a resource is created. However, these are not the types of
+systems generally encouraged, and a client **should not** retry these types of
+errors.
+
+These codes generally **should not** be retried for any request:
+
+- `RESOURCE_EXHAUSTED`: This may be a signal that the user's quota is exhausted.
+  Retries therefore may not be expected to work for several hours; meanwhile the
+  retries may have billing implications. If `RESOURCE_EXHAUSTED` is used for
+  other reasons than quota and the expected time for the resource to become
+  available is much shorter, it may be retryable.
+- `NOT_FOUND`: The resource may be created in the future.
+- `ALREADY_EXISTS`: The resource may be deleted in the future.
+- `PERMISSION_DENIED`: The user may be authenticated in the future.
+- `FAILED_PRECONDITION`: The system state may change in the future.
+- `OUT_OF_RANGE`: The range may be extended in the future.
+- `UNIMPLEMENTED`: The RPC may be implemented in the future.
+- `UNKNOWN`: `UNKNOWN` and `INTERNAL` are both generic error codes - they are
+  errors that may be generated from a wide breadth of causes. Known services
+  within the system **should** generate `INTERNAL` errors when the error doesn't
+  fit any other code. `UNKNOWN` errors are reserved for truly
+  unknown-to-the-system errors, and therefore **should** immediately surfaced to
+  the user.
+
+## Appendix
+
+### Idempotency and pragmaticism
+
+A common concern with retrying `CREATE` and `DELETE` operations is that they are
+non-idempotent operations. The functional concern can be summarized with this
+example:
+
+```
+Get("Foo")
+> NOT_FOUND
+Create("Foo") // Let's work under the assumption that this should be retried.
+// ----> UNAVAILABLE
+// ----> Create("Foo") # retry #1
+// ----> UNAVAILABLE   # actually, the request succeeded, but the connection died during the response
+// ----> Create("Foo") # retry #2
+// ----> ALREADY_EXISTS
+> ALREADY_EXISTS
+```
+
+Here, an unexpected result - `ALREADY_EXISTS` - is returned from a Create. The
+resource didn't exist before the call was initiated, so the user may wonder,
+"How did I receive this result?".
+
+Does this unexpected result make the call non-idempotent?
+
+The definition of idempotency is unfortunately vague in this regard. [Oxford
```

Comment:
I'm not sure this is helpful - after all, it's *a* definition, but there could be multiple definitions, some of which are less vague.

Reply:
Acknowledged, but I think that a definition is better than no definition. It's an attempt at starting at a reasonable foundation before continuing with rationale.

I could rephrase this to, "One definition of idempotency is [...]". Would that work, do you think?

Reply:
Considering that AIPs are specs, we can _stipulate_ what our definition of idempotency is for the purpose of this document.